Refactor

clarkfitzg opened the pull request:
It's going to be much easier and more maintainable to write drivers if the internals in ddR are cleaned up a bit. So that's what I've been working on the last few days, as well as [documentation](https://github.com/vertica/ddR/wiki/Drivers-and-Backends).

My goal with this first pass is to not change anything too drastically, just tighten up what's already there. The biggest change is in the initialization for the parallel driver, and relying a bit more on the parallel library.

Right now one test is failing, but I'll likely have that fixed tomorrow and be ready for a code review. cc @etduwx @fun-indra @lawremi

clarkfitzg commented on ddR/R/pdriver.R:
```diff
@@ -25,63 +27,74 @@ setClass("ParallelddR", contains="ddRDriver")
+
+    # Normalize executors to positive integer
+    if(executors == "all"){
+        # NIT- Figure out how to suppress this message:
```
Was unsuccessful trying [various things](https://github.com/clarkfitzg/junkyard/blob/master/cat_output.R) with `capture.output` and `sink` to suppress stderr message from system call to `cat`. Looking at the code for `parallel::detectCores` it should be ok to just use default `all.tests=FALSE`.

clarkfitzg commented on ddR/R/ddR.R:
```diff
@@ -80,32 +80,40 @@ useBackend <- function(driver, ...) {
+#'      change backends
 #' @export
-setGeneric("init", function(x,...) standardGeneric("init"))
+setGeneric("init", function(driver,...) standardGeneric("init"))
```
Is there a convention in R for this argument name? Something like `self` in Python?

lawremi replied:
`x` is the convention I always use.

clarkfitzg replied:
I'll change these back to `x` then.

clarkfitzg replied:
@lawremi exactly when does this convention apply? Looking at `do_mapply` and `combine` the first arg is `driver`, which seems like a nice descriptive name. So I was changing `init` and others to match. For internal consistency they should either all be `x` or `driver`.

lawremi replied:
Many generics in R use the `x` convention. While it would be fine to use `driver` for the generics we define, there is going to be inconsistency with other generics. 

A related problem is sharing generics with other popular packages. For example, Bioconductor defines the `combine(x, y, ...)` generic. We should try to avoid conflict if at all possible.

clarkfitzg replied:
good to know. Then inside ddR they will stay a mix of `x` and `driver`.

clarkfitzg commented on ddR/R/zzz.R:
```diff
@@ -16,5 +16,12 @@
+
+    # This should run if the default of parallel is being used as the
+    # driver.
+    # Maybe there's a better place to put this?
```
Open to suggestions on where to initialize the default backend.

etduwx replied:
What's wrong with how it was originally done, on line 29 of pdriver.R, where the env variable is set directly?

clarkfitzg replied:
1. Putting it here makes it more explicit that we're choosing parallel as the default.
2. `pdriver` is a model for other drivers, so it should avoid changing package level variables if possible.
3. There is some overhead to set up a parallel cluster, and this is a step in the direction of avoiding that overhead.

clarkfitzg commented:
Tests all pass; everything seems to work. Ready for code review now.

clarkfitzg commented on ddR/R/pdriver.R:
```diff
@@ -15,8 +15,14 @@
 setClass("ParallelddR", contains="ddRDriver")
 
+# TODO Clark: Change from "parallel" to more descriptive name like
+# "parallel.driver"
```
I'd like to know how open everyone is to breaking changes like this.

etduwx commented on ddR/R/ddR.R:
```diff
@@ -55,20 +56,19 @@ ddR.env$nexecutors <- 1
-  if(!extends(class(driver)[[1]],"ddRDriver")) stop("Invalid driver object specified")
+    if(!is.null(ddR.env$driver)) shutdown(ddR.env$driver)
 
-  if(!extends(driver@DListClass,"DObject")) stop("The driver DList class does not extend ddR::Dobject")
```
The point of these checks is to protect the user from supplying an invalid driver object, i.e., he/she may type 
`useBackend(foo)`, where `foo` may be anything. We need not only check for inheritance of the driver object but also the dobject types it uses.

 I'm not opposed to having these checks in the tests, but I think that they should still be in the package code.

clarkfitzg replied:
Has a need for this come up in the past?

This approach is simpler, and the only way it can fail is if people are writing their own drivers. In this case they probably expect to encounter problems, and should only be hopeful that it will work when it passes the new `test-driver-common.R` file. Even when ddR is fully mature my guess is there will be ~5 robust drivers that people use. 

There are many possible checks that could be done to make sure the driver _actually_ works- I would think that checking for the definition of `do_mapply` and `do_collect` would be the most important.

etduwx commented on ddR/R/ddR.R:
```diff
@@ -246,6 +255,8 @@ dmapply <- function(FUN ,..., MoreArgs=list(),
   # Ensure that ... arguments are of equal length. length() works correctly for data.frame,
   # arrays, and lists
   lens <- vapply(dargs,function(x) {
+    # TODO Clark: Why even allow the possibility of this happening where
```
How would you prevent it from happening? If the user creates a dobject with one driver, e.g.:

`useBackend(parallel)`
`a <- dlapply(...)`
`useBackend(Spark)`
`b <- dlapply(...)`

We need to catch the error if/when the user does
`dmapply(function()...,a,b)`, since `a` and `b` came from different backends. All input objects need to have been derived from the same backend.

clarkfitzg replied:
ddR uses a global backend, so I'm not seeing how it would work to be using and switching between multiple backends in the same session.

The solution I had in mind is to allow only one backend per session (implemented inside `useBackend`), but maybe this is too restrictive?

fun-indra replied:
Does session mean R session? How would you support the following case: I start R,  run an algorithm on a small dataset using parallel, and then use Spark with ddR for a larger dataset. We should not force someone to restart the R session.

clarkfitzg replied:
Yes, I was thinking of requiring them to restart the R session. @fun-indra if I was doing that I'd definitely be restarting my R session. But not everyone has that workflow, and I agree it's better not to force it on people.

clarkfitzg commented on ddR/tests/testthat/test-pdriver.R:
```diff
@@ -0,0 +1,56 @@
+
+    # TODO Clark: Currently this works, but maybe better to throw an
+    # error if mixing clusters like this.
+    out <- collect(dmapply(c, dl2, dl3))
```
Re previous discussion, should this be raising an error?

etduwx replied:
@fun-indra Is the above okay?

fun-indra replied:
Yes, we should throw an error.

If ddR evolves substantially, we can imagine a scenario where one of dl2 or dl3 will be pulled from it's current backend and moved to the other backend for execution, or  dmapply itself is distributed across the two backends (similar to concepts in federated databases). But we are not there, and it's better to throw an error.

clarkfitzg replied:
Ok, I'll modify it to throw an error. The way I plan on implementing this is to have references to current running backends. Something like this:

```
p2 <- useBackend("parallel", executors = 2)
p3 <- useBackend("parallel", executors = 3) # this is now set as the global one
shutdown(p2)
```

clarkfitzg commented:
I'm going to refrain from making any more changes in this PR- just addressing your feedback now. Which is helpful, thanks! 👍

clarkfitzg commented:
Clarifying- I'll make any changes necessary to get this PR into master so I can build from it. @etduwx @fun-indra can you take a look and tell me what it needs?

fun-indra commented:
Looks reasonable to me. We can merge. @etduwx any reservations?

etduwx commented on ddR/NAMESPACE:
```diff
@@ -40,7 +39,7 @@ export(is.dlist)
 export(is.sparse_darray)
 export(nparts)
-export(parallel)
+export(parallel.ddR)
```
I don't really have any major objection to changing the name to this, but we have to agree to consistency. If we follow this convention, this means that backends should always be referred to with a ".ddR" suffix, e.g., one would have to type `useBackend(distributedR.ddR)` for Distributed R. I think this is slightly more cumbersome to type than just `useBackend(distributedR)`. Thoughts?

fun-indra replied:
I agree the .ddR looks more cumbersome.

lawremi replied:
Right now there is a redundancy when using `library()` to load a backend package and then `useBackend()` to actually enable it. Maybe `useBackend()` could take the backend name as a string, and the function would automatically find the corresponding package (if external), load it, and use its backend. That's a better abstraction than requiring the user to specify the backend package name, and the object name. We could also have an `availableBackends()` that lists possible values. I wonder if there are any facilities (e.g., with RStudio) so that it offers completions like `library()`. Another idea is to have an object like `ddr` that is a dynamic list-like database of backends, along with a `use()` generic, so we could have `use(ddr$parallel)`. Just throwing out some ideas.

clarkfitzg replied:
Let's see how the options look together:

1. `library(Spark.ddR); useBackend(Spark.ddR, master = "local", ...)` Current PR
1. `useBackend("spark", master = "local", ...)`
2. `use(ddR$spark, master = "local", ...)`

The 2nd option is appealing, along with the functionality @lawremi mentions. It seems like a clear and concise option.

etduwx commented:
Seems ok. Just added a couple of more minor comments.

clarkfitzg commented:
To enable correct switching of backends I changed the distributed objects so that they reference an instance of a backend driver that created them. This required a change to the OO model of the drivers. I updated [the wiki](https://github.com/vertica/ddR/wiki/Drivers-and-Backends) to reflect what's here- feel free to take a look and give me any feedback.

clarkfitzg commented:
BTW with these changes will need a bit more testing and documentation. I'll comment when I feel it's ready, probably tomorrow.

clarkfitzg commented:
The current state looks good to me.

- Feedback has been incorporated
- ddR tests all pass
- Algorithm tests all pass
- Examples run with same issues as before

There are a couple related things left to be done:

- Modify `distributedR.ddR` to fit with the new OO driver / backend model. This shouldn't be difficult, but it will require a running distributedR cluster to test it, which I don't have on hand.
- Make sure docs build properly

@fun-indra @etduwx, the ball is back in your court. Don't feel shy about feedback, I'm happy to continue iterating on this.

etduwx commented:
@clarkfitzg ,

Thanks for your work on this, and I think things look to be in good shape. However, my one reservation about merging this code now is the fact that it may break Distributed R compatibility, so I suggest we wait until that's fixed and tested before we do so. 

Let us know if you'd like to try out DistR and need help setting it up. If not, I'll probably also eventually get around to testing it with your new code.

clarkfitzg commented:
Ok I'll make the changes and have a go setting up DistR on my department servers and running it there. Will let you know if I get stuck.

clarkfitzg commented:
Made changes to distributedR.ddR, but it definitely need to be tested.

I'm not able to set up DistR since I don't have sudo access to dept servers. Any chance HP has a running cluster that I can use? Otherwise running these tests is going to be up to @etduwx.

clarkfitzg commented:
Multiple physical machines on a SNOW cluster pass the tests. MPI also works. I didn't try an NWS (NetWorkSpaces) cluster because it's a fair amount of configuration, and it looks like the project hasn't been actively developed in the last few years. Still I think NWS has the potential to do some very cool things using `parts` with ddR.

clarkfitzg commented on ddR/R/pdriver.R:
```diff
@@ -1,90 +1,125 @@
+#'
+#' # PSOCK cluster on multiple machines (assuming you can ssh user@server1):
+#' servers <- c("server1", "server2")
+#' useBackend("parallel", servers, type = "PSOCK")
```
@dselivanov This is how one would use multiple physical machines.

dselivanov replied:
Imagine I have 2 machines with 4 cores on each. How to create "PSOCK" cluster with 8 workers?

HenrikBengtsson replied:
Machine names can be replicated, e.g. rep(c("server1", "server2"), each=4)
to get 8 workers in total.

On Sep 4, 2016 22:43, "Dmitriy Selivanov" <notifications@github.com> wrote:
>
> In ddR/R/pdriver.R:
>
> > #' @examples > #' \dontrun{ > -#' useBackend(parallel,executors=4) >
+#' # Cluster on 2 cores with default type: > +#' useBackend("parallel", 2)
> +#' > +#' # PSOCK cluster on multiple machines (assuming you can ssh
user@server1): > +#' servers <- c("server1", "server2") > +#'
useBackend("parallel", servers, type = "PSOCK")
>
> Imagine I have 2 machines with 4 cores on each. How to create "PSOCK"
cluster with 8 workers?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

clarkfitzg commented:
Thanks for the feedback @dselivanov @HenrikBengtsson - I added your suggestions into the docs.

clarkfitzg commented:
After running ddR on a few GB size data sets I've observed that `parallel::mcmapply` and `parallel::clusterMap` on a `FORK` cluster do completely different things, with `mcmapply` being much more efficient. This surprised me.

Looking back at it, the implementation of the `Spark` driver is basically the same as for a `SNOW` cluster, while `FORK` is a different from both of these. (Assuming we're trying to get distributed data to live on the nodes of a SNOW cluster).

clarkfitzg commented on ddR/R/ddR.R:
```diff
@@ -178,16 +219,202 @@ setGeneric("do_collect", function(x, parts) standardGeneric("do_collect"))
+
+# TODO Clark: The line below needs to change if data is actually
+# to remain distributed
+      dots[[num]] <- collect(dots[[num]])
```
I believe this line is the culprit for excessive memory consumption with SNOW, and it should change to support truly distributed data. In general the `do_dmapply` should handle distributed objects. Leaving this for now because it works and it's useful for testing purposes.
